allennlp/data/dataset_readers/dataset_reader.py

```diff
@@ -157,9 +173,29 @@ def _read(self, file_path: str) -> Iterable[Instance]:
         raise NotImplementedError
 
     def _instances_from_cache_file(self, cache_filename: str) -> Iterable[Instance]:
-        with open(cache_filename, 'r') as cache_file:
-            for line in cache_file:
-                yield self.deserialize_instance(line.strip())
+        if self._cache_method == 'jsonpickle':
+            with open(cache_filename, 'r') as cache_file:
+                for line in cache_file:
+                    yield self.deserialize_instance(line.strip())
+        elif self._cache_method == 'pickle':
+            # Need to yield the entries one at a time do provide a consistent interface.
+            with open(cache_filename, 'rb') as cache_file:
+                cached = pkl.load(cache_file)
+                for entry in cached:
+                    yield entry
+        else:
+            raise ConfigurationError('Available caching methods are `jsonpickle` and `pickle`.')
+
+    def _instances_to_cache_file(self, cache_file, instances):
```

Comment:
Here's a potential solution that would make it so you don't need to add the `cache_method` flag, or the stringly-typed argument in various places.  Just add this method, and this method only, to `DatasetReader`, and override it in your `DatasetReader` to use pickle instead of jsonpickle.  This kind of overriding was the original intent of the design here, we just assumed that everything would be line-by-line, so we could handle lazy datasets also.  I think it's fine to add this overridable method, and let people change it to not line-by-line if they want, with the understanding that it'll break lazy caching if they want to do that.